jib-core/src/main/java/com/google/cloud/tools/jib/tar/TarExtractor.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2019 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.tar;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
+
+/** Extracts a tarball. */
+public class TarExtractor {
+
+  /**
+   * Extracts a tarball to the specified destination.
+   *
+   * @param source the tarball to extract
+   * @param destination the output directory
+   * @throws IOException if extraction fails
+   */
+  public static void extract(Path source, Path destination) throws IOException {
```

Comment:
Now I think what I missed. Ideally, we will have better performance by accepting an input stream, because `docker save` can output to stdout. I see the plan was to docker-save to a tarball and read it again to extract, but we can avoid reading twice and having to do disk write, saving space and time. (Of course, the assumption is that we don't need the tarball.) I think it's worth avoiding extra read-write on disk.

Reply:
We don't really "need" the tarball, but it lets us reuse the code path for using a tarball base image. I'm imagining adding 2 new build steps (a docker save build step that returns a path to a tarball, and an extraction build step that converts the tar into the `Image` + `ContainerConfigurationTemplate` + layers). A docker daemon base image would use both new build steps, and a tarball base image would only need to use the second one.

You are right about performance, although it might be tricky navigating the tar stream in a non-linear fashion, wouldn't it? We need to parse the manifest first to get the names of the layers + configuration, which means we'd need to skip over entries if the manifest isn't the first entry in the tar. Unless we could determine which files are which based on extension, or if we can traverse the tar contents out of order? I'm not too familiar with the tar archive api.

Reply:
Okay, let's just move forward with this. I will have a chance to revisit this once your design is implemented. I don't think I can suggest something concrete until I have a firm understanding of the implementation.